msteams-platform/messaging-extensions/build-api-based-message-extension.md

```diff
@@ -0,0 +1,462 @@
+                            ]
+                        },
+                        {
+                            "type": "FactSet",
+                            "facts": [
+                                {
+                                    "title": "Repair ID:",
+                                    "value": "${if(id, id, 'N/A')}"
+                                },
+                                {
+                                    "title": "Date:",
+                                    "value": "${if(date, date, 'N/A')}"
+                                }
+                            ]
+                        }
+                    ]
+                }
+            ]
+        },
+        "previewCardTemplate": {
+            "title": "Title: ${if(title, title, 'N/A')}",
+            "subtitle": "Description: ${if(description, description, 'N/A')}",
+            "text": "Assigned To: ${if(assignedTo, assignedTo, 'N/A')}",
+            "image": {
+                "url": "${if(image, image, '')}"
+            }
+        }
+    }
+    ```
+
+    </details>
+
+1. Update app manifest: Update the app manifest with the composeExtensions property. The following is an example of the app manifest with the `composeExtensions` property:
+
+   ```json
+      "composeExtensions": [
+        {
+          "type": "ApiBased",
+          "apiSpecFile": "listrepairsapispec.yaml",
+          "commands": [
+            {
+              "context": [
+                "compose"
+              ],
+              "type": "query",
+              "id": "listRepairs",
+              "title": "List repairs",
+              "parameters": [
+                {
+                  "title": "Filter",
+                  "name": "filter",
+                  "description": "Filter repairs by who they're assigned to."
+                }
+              ],
+              "apiResponseRenderingTemplateFile": "listrepairsresponsetemplate.json"
+            }
+          ]
+        }
+      ],
+      "validDomains": [
+        "repairs-api-2023.azurewebsites.net"
+      ]
+    }
+
+   ```
+
+   Ensure that the app has the following information:
+
+    * App name
+    * App color icon
+    * App outline icon
+    * Terms of Service URL
+    * Privacy policy URL
+    * The authentication type in `composeExtension.auth`: no auth, aad, oauth, or service_http.
+
+   **Query Parameters**
+ 
+   |Name  |Description  |
+   |---------|---------|
+   |`composeExtension.type`     |   We take an OpenAPI spec and it can either be used by BizChat to figure out how/when to call the API, or used by dev toolkit to help generate manifest/Adaptive Card templates.  Update the value as `ApiBased`. |
+   |`composeExtension.apiSpecificationFile`     |  This references an OpenAPI spec file in the app package. Used to get the endpoint url, if not specified, construct the http request and also used by Biz chat to figure out how to call and process a response.  Include when type is `ApiBased`.      |
```

Comment:
Similar question here--how much is implementation detail? Can/should this be simplified?